### PR TITLE
[Feature] Adds utils method isin

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

A similar method to torch.isin but for TensorDicts.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/tensordict/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/tensordict/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: tensordict/utils.py [hunk 1]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
         logging.info(indent + os.path.basename(path))
 
 
+def isin(
```
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/utils.py:1854]: Maybe rename in `contains`?

File: tensordict/utils.py [hunk 2]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+    tensordict: TensorDictBase,
+    reference_tensordict: TensorDictBase,
+    key: str,
+    dim: int = 0,
+    invert: bool = False,
+):
```
Comment (vmoens) [inline tensordict/utils.py:1860]: can we type the output too?

File: tensordict/utils.py [hunk 3]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+    """Tests if each element of 'key' in tensordict 'dim' is also present in the reference_tensordict.
+
+    This function returns a boolean tensor of length  ´tensordict.batch_size[dim]´ that is True for elements in
+    'key' that are also present in the reference_tensordict. This function assumes that both tensordict and
+    reference_tensordict have the same batch size and contain the specified key, otherwise it will raise an error.
```
Comment (vmoens) [inline tensordict/utils.py:1865]:
```suggestion
    """Tests if each element of ``key`` in tensordict ``dim`` is also present in the reference_tensordict.

    This function returns a boolean tensor of length  ``input.batch_size[dim]`` that is ``True`` for elements in
    the entry ``key`` that are also present in the ``reference``. This function assumes that both ``input`` and
    ``reference`` have the same batch size and contain the specified entry, otherwise an error will be raised.
```

File: tensordict/utils.py [hunk 4]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+
+    Args:
+        tensordict (TensorDictBase): Input TensorDict.
+        reference_tensordict (TensorDictBase): Target TensorDict against which to test.
+        key (str): The key to test.
+        dim (int): The dimension along which to test. Default: 0
+        invert (bool):  If True, inverts the boolean return tensor, resulting in True values for elements
+            not in reference_tensordict. Default: False
```
Comment (vmoens) [inline tensordict/utils.py:1873]:
```suggestion
    Args:
        input (TensorDictBase): Input TensorDict.
        reference (TensorDictBase): Target TensorDict to compare to ``input``.
        key (NestedKey): The key to test.
        dim (int, optional): The dimension along which to test. Defaults to ``0``.
    
    Keyword Args:
        invert (bool, optional):  If ``True``, inverts the boolean return tensor, resulting in ``True`` values for elements
            not in ``reference``. Defaults to ``False``.
```

To me `invert` brings very little and should be removed. You could just to `~isin(td0, td1)` and it would give the same result for just 1 extra character. I tend to think about kwargs as bringing non-trivial functionalities.

File: tensordict/utils.py [hunk 5]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+
+    Returns:
+        out (Tensor): A boolean tensor of length ´tensordict.batch_size[dim]´ that is True for elements in
+            'key' that are also present in the reference_tensordict or False if invert is True.
```
Comment (vmoens) [inline tensordict/utils.py:1877]:
```suggestion
        out (Tensor): A boolean tensor of length ``input.batch_size[dim]`` that is ``True`` for elements in
            ``key`` that are also present in the ``reference`` or ``False`` if invert is ``True``.
```

File: tensordict/utils.py [hunk 6]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+        >>> td_ref = TensorDict(
+        ...     {
+        ...         "tensor1": torch.tensor([[1, 2, 3], [4, 5, 6], [10, 11, 12]]),
+        ...         "tensor2": torch.tensor([[10, 20], [30, 40], [50, 60]]),
+        ...     },
+        ...     batch_size=[3],
+        ... )
+        >>> in_reference = is_in_reference(td, td_ref, key="tensor1")
+        >>> expected_in_reference = torch.tensor([True, True, True, False])
+        >>> assert torch.equal(in_reference, expected_in_reference)
```
Comment (vmoens) [inline tensordict/utils.py:1896]: there's also torch.testing.assert_close IIRC

File: tensordict/utils.py [hunk 7]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+    """
+    # Check key is present in both tensordict and reference_tensordict
+    if key not in tensordict.keys(include_nested=True):
+        raise KeyError(f"Key '{key}' not found in tensordict.")
+    if key not in reference_tensordict.keys(include_nested=True):
+        raise KeyError(f"Key '{key}' not found in reference_tensordict.")
```
Comment (vmoens) [inline tensordict/utils.py:1902]: What I usually do (to check if the key is there only once and make things faster)
```suggestion
    reference_tensor = reference_tensordict.get(key, default=None)
    target_tensor = tensordict.get(key, default=None)

    if not isinstance(target_tensor, torch.Tensor):
        raise KeyError(f"Key '{key}' not found in tensordict or not a tensor.")
    if not isinstance(reference_tensor, torch.Tensor):
        raise KeyError(f"Key '{key}' not found in reference_tensordict or not a tensor.")
```

File: tensordict/utils.py [hunk 8]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+
+    # Check that both TensorDicts have the same number of dimensions
+    if len(tensordict.batch_size) != len(reference_tensordict.batch_size):
+        raise ValueError(
+            "The number of dimensions in the batch size of the tensordict and reference_tensordict must be the same."
+        )
+
+    # Check that dim is valid
+    if dim >= len(tensordict.batch_size):
```
Comment (vmoens) [inline tensordict/utils.py:1911]: let's make it work with neg dims too

File: tensordict/utils.py [hunk 9]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+    if key not in tensordict.keys(include_nested=True):
+        raise KeyError(f"Key '{key}' not found in tensordict.")
+    if key not in reference_tensordict.keys(include_nested=True):
+        raise KeyError(f"Key '{key}' not found in reference_tensordict.")
+
+    # Check that both TensorDicts have the same number of dimensions
+    if len(tensordict.batch_size) != len(reference_tensordict.batch_size):
+        raise ValueError(
+            "The number of dimensions in the batch size of the tensordict and reference_tensordict must be the same."
+        )
```
Comment (vmoens) [inline tensordict/utils.py:1908]: one should be greater than the other no?
like `tensordict.batch_size[dim] > reference_tensordict.batch_size[dim]`

File: tensordict/utils.py [hunk 10]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+
+    # Get the data
+    reference_tensor = reference_tensordict.get(key)
+    target_tensor = tensordict.get(key)
+    N = reference_tensor.shape[0]
+
+    # Find the common indices
+    cat_data = torch.cat([reference_tensor, target_tensor], dim=0)
+    _, unique_indices = torch.unique(cat_data, dim=0, sorted=True, return_inverse=True)
+    out = torch.isin(unique_indices[N:], unique_indices[:N], assume_unique=True)
```
Comment (vmoens) [inline tensordict/utils.py:1924]: I don't see where dim is used

**Review by vmoens — COMMENTED** [review]
Thanks, great feature! I left some comments

File: tensordict/utils.py (hunk 9 above)
Comment (albertbou92) [reply tensordict/utils.py]: the only important thing is that both TD have the same number of dimensions. How many elements are present in each dimension does not really matter. So we only check that `len(tensordict.batch_size) != len(reference_tensordict.batch_size)`

**Comment by vmoens** [comment]
RE renaming in contains I think isin isn't too bad after all since it's the PT name

**Comment by albertbou92** [comment]
> RE renaming in contains I think isin isn't too bad after all since it's the PT name

haha this is why I named it like that. I am fine with both.

File: tensordict/utils.py [hunk 11]
```diff
@@ -1851,6 +1851,85 @@ def format_size(size):
...
+
+    This function returns a boolean tensor of length  ``input.batch_size[dim]`` that is ``True`` for elements in
+    the entry ``key`` that are also present in the ``reference``. This function assumes that both ``input`` and
+    ``reference`` have the same batch size and contain the specified entry, otherwise an error will be raised.
+
+    Args:
+        input (TensorDictBase): Input TensorDict.
+        reference (TensorDictBase): Target TensorDict against which to test.
+        key (Nestedkey): The key to test.
+        dim (int, optionl): The dimension along which to test. Default: 0
```
Comment (vmoens) [inline tensordict/utils.py:1870]:
```suggestion
        dim (int, optional): The dimension along which to test. Defaults to ``0``.
```

**Review by vmoens — COMMENTED** [review]
LGTM, just 2 small changes

**Review by vmoens — APPROVED** [review]
